### PR TITLE
test(halo/evmengine): prepare proposal tests

### PR DESCRIPTION
**Pull request by Lazar955**
Adds unit tests for prepare proposal

task: https://app.asana.com/0/1206208509925075/1206657733675480/

### CODE REVIEW

File: lib/ethclient/enginemock.go [hunk 1]
```diff
@@ -139,7 +139,7 @@ func (m *engineMock) ForkchoiceUpdatedV2(ctx context.Context, update engine.Fork
 
 	// If we have payload attributes, make a new payload
 	if attrs != nil {
-		payload, err := makePayload(m.fuzzer, m.head.NumberU64()+1, attrs.Timestamp, update.HeadBlockHash)
+		payload, err := makePayload(m.fuzzer, m.head.NumberU64()+1, attrs.Timestamp, update.HeadBlockHash, attrs.SuggestedFeeRecipient)
```
Comment (corverroos) [inline lib/ethclient/enginemock.go:142]: rather pass in attrs.Randao as well

File: halo/evmengine/keeper/abci_internal_test.go [hunk 2]
```diff
@@ -0,0 +1,544 @@
...
+		ap := MockAddressProvider{}
+		keeper := NewKeeper(cdc, storeService, &mockEngine, txConfig, ap)
+
+		ts := time.Now()
+		latestHeight, err := mockEngine.BlockNumber(ctx)
+		require.NoError(t, err)
+		latestBlock, err := mockEngine.BlockByNumber(ctx, big.NewInt(int64(latestHeight)))
+		require.NoError(t, err)
+
+		mockEngine.pushPayload(t, ctx, ap.LocalAddress(), latestBlock.Hash(), ts)
```
Comment (corverroos) [inline halo/evmengine/keeper/abci_internal_test.go:213]: add much more comments, one for every step, it really isn't very clear what is happening here.

File: halo/evmengine/keeper/abci_internal_test.go [hunk 3]
```diff
@@ -0,0 +1,544 @@
...
+	t.Helper()
+	key := storetypes.NewKVStoreKey("test")
+	storeService := runtime.NewKVStoreService(key)
+	testCtx := testutil.DefaultContextWithDB(t, key, storetypes.NewTransientStoreKey("transient_test"))
+	ctx := testCtx.Ctx.WithBlockHeader(cmtproto.Header{Time: cmttime.Now()})
+
+	return ctx, storeService
+}
+
+func getCodec(t *testing.T) *codec.ProtoCodec {
```
Comment (corverroos) [inline halo/evmengine/keeper/abci_internal_test.go:267]: What about rather:
```
moduletestutil "github.com/cosmos/cosmos-sdk/types/module/testutil"
...
moduletestutil.MakeTestEncodingConfig().Codec

```

File: halo/evmengine/keeper/abci_internal_test.go [hunk 4]
```diff
@@ -0,0 +1,544 @@
...
+	// TODO implement me
+	panic("implement me")
+}
+
+func (m *MockEngineAPI) EstimateGas(ctx context.Context, call ethereum.CallMsg) (uint64, error) {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (m *MockEngineAPI) SuggestGasPrice(ctx context.Context) (*big.Int, error) {
```
Comment (corverroos) [inline halo/evmengine/keeper/abci_internal_test.go:373]: can remove implementme functions, just embed the interface you want to implement in the struct itself

File: halo/evmengine/keeper/abci_internal_test.go [hunk 5]
```diff
@@ -0,0 +1,544 @@
...
+		BeaconRoot:            nil,
+	}
+
+	resp, err := m.ForkchoiceUpdatedV2(ctx, state, &payloadAttrs)
+	require.NoError(t, err)
+
+	return resp.PayloadID
+}
+
+func (m *MockEngineAPI) nextBlock(t *testing.T, height uint64, timestamp uint64, parentHash common.Hash,
```
Comment (corverroos) [inline halo/evmengine/keeper/abci_internal_test.go:521]: add some comments

File: halo/evmengine/keeper/abci_internal_test.go [hunk 6]
```diff
@@ -0,0 +1,544 @@
...
+
+func (m *MockEngineAPI) GetPayloadV2(ctx context.Context, payloadID eengine.PayloadID) (*eengine.ExecutionPayloadEnvelope, error) {
+	return m.Mock.GetPayloadV2(ctx, payloadID)
+}
+
+func (m *MockEngineAPI) GetPayloadV3(ctx context.Context, payloadID eengine.PayloadID) (*eengine.ExecutionPayloadEnvelope, error) {
+	panic("implement me")
+}
+
+func (m *MockEngineAPI) pushPayload(t *testing.T, ctx context.Context, feeRecipient common.Address, blockHash common.Hash, ts time.Time) *eengine.PayloadID {
```
Comment (corverroos) [inline halo/evmengine/keeper/abci_internal_test.go:499]: comments

File: halo/evmengine/keeper/abci_internal_test.go [hunk 7]
```diff
@@ -0,0 +1,544 @@
...
+		ctx, storeService := setupCtxStore(t)
+		cdc := getCodec(t)
+		txConfig := authtx.NewTxConfig(cdc, nil)
+
+		me, err := ethclient.NewEngineMock()
+		require.NoError(t, err)
+		mockEngine := MockEngineAPI{
+			Mock:   me,
+			fuzzer: ethclient.NewFuzzer(time.Now().Truncate(time.Hour * 24).Unix()),
+		}
```
Comment (corverroos) [inline halo/evmengine/keeper/abci_internal_test.go:203]: make a function `newEngineAPIMock`

**Review by corverroos — APPROVED** [review]

File: halo/evmengine/keeper/abci_internal_test.go (hunk 3 above)
Comment (Lazar955) [reply halo/evmengine/keeper/abci_internal_test.go]: Doesn't help that much as we still have to register all the interfaces for the codec

**Pull request merged**